infrastructures/openshift/src/main/java/org/eclipse/che/workspace/infrastructure/openshift/project/OpenShiftProjectFactory.java

```diff
@@ -30,19 +46,86 @@
  */
 @Singleton
 public class OpenShiftProjectFactory extends KubernetesNamespaceFactory {
+  private static final Logger LOG = LoggerFactory.getLogger(OpenShiftProjectFactory.class);
 
   private final OpenShiftClientFactory clientFactory;
+  private final String defaultNamespaceName;
+  private final boolean allowUserDefinedNamespaces;
 
   @Inject
   public OpenShiftProjectFactory(
       @Nullable @Named("che.infra.openshift.project") String projectName,
       @Nullable @Named("che.infra.kubernetes.service_account_name") String serviceAccountName,
       @Nullable @Named("che.infra.kubernetes.cluster_role_name") String clusterRoleName,
-      OpenShiftClientFactory clientFactory) {
-    super(projectName, serviceAccountName, clusterRoleName, clientFactory);
+      @Named("che.infra.kubernetes.namespace.default") String defaultNamespaceName,
+      @Named("che.infra.kubernetes.namespace.allow_user_defined")
+          boolean allowUserDefinedNamespaces,
+      OpenShiftClientFactory clientFactory,
+      OpenShiftClientConfigFactory clientConfigFactory) {
+    super(
+        projectName,
+        serviceAccountName,
+        clusterRoleName,
+        defaultNamespaceName,
+        allowUserDefinedNamespaces,
+        clientFactory);
+    if (allowUserDefinedNamespaces && !clientConfigFactory.isPersonalized()) {
+      LOG.warn(
+          "Users allowed to list projects but all reuse Che Server configured service account. "
+              + "Consider configuring OpenShift OAuth to personalize credentials that will be used for cluster access.");
+    }
+    this.defaultNamespaceName = defaultNamespaceName;
+    this.allowUserDefinedNamespaces = allowUserDefinedNamespaces;
     this.clientFactory = clientFactory;
   }
 
+  @Override
+  public List<KubernetesNamespaceMeta> list() throws InfrastructureException {
+    if (!allowUserDefinedNamespaces) {
+      // return only default project if user defined are not allowed
+      String evaluatedName = evalDefaultNamespaceName(EnvironmentContext.getCurrent().getSubject());
+
+      Project project = clientFactory.createOC().projects().withName(evaluatedName).get();
+
+      KubernetesNamespaceMeta defaultNamespace;
+      if (project == null) {
+        // if the predefined project does not exist - return dummy info and it will be created
+        // during the first workspace start
+        defaultNamespace = new KubernetesNamespaceMetaImpl(evaluatedName);
+      } else {
+        defaultNamespace = asNamespaceMeta(project);
+      }
+
+      defaultNamespace.getAttributes().put(DEFAULT_ATTRIBUTE, "true");
+      return singletonList(defaultNamespace);
+    }
```

Comment:
Same as for `KubernetesNamespaceFactory`, consider extracting this branch to a separate method.